File: advent-of-code/2015/05/02/solution.cs
using System;
using System.IO;
using System.Linq;

bool IsNice(string s)
{
    // Check for a pair that appears at least twice without overlapping
    bool hasPair = false;
    for (int i = 0; i < s.Length - 1; i++)
    {
        string pair = s.Substring(i, 2);
        // Look for the same pair starting at least 2 positions later
        if (s.IndexOf(pair, i + 2) >= 0)
        {
            hasPair = true;
            break;
        }
    }
    if (!hasPair)
    {
        return false;
    }

    // Check for a letter that repeats with exactly one letter between
    bool hasRepeat = false;
    for (int i = 0; i < s.Length - 2; i++)
    {
        if (s[i] == s[i + 2])
        {
            hasRepeat = true;
            break;
        }
    }

    return hasRepeat;
}

// Test cases with assertions
System.Diagnostics.Debug.Assert(IsNice("qjhvhtzxzqqjkmpb") == true);
System.Diagnostics.Debug.Assert(IsNice("xxyxx") == true);
System.Diagnostics.Debug.Assert(IsNice("uurcxstgmygtbstg") == false);
System.Diagnostics.Debug.Assert(IsNice("ieodomkazucvgmuy") == false);

// Read input and calculate result
string inputPath = Path.Combine("..", "input.txt");
string[] strings = File.ReadAllLines(inputPath);

int result = strings.Count(IsNice);
Console.WriteLine($"C#: {result}");

File: advent-of-code/2015/05/02/solution.rs
use std::fs;

fn is_nice(s: &str) -> bool {
    // Check for a pair that appears at least twice without overlapping
    let has_pair = (0..s.len()-1).any(|i| {
        let pair = &s[i..i+2];
        s[i+2..].contains(pair)
    });
    
    if !has_pair {
        return false;
    }

    // Check for a letter that repeats with exactly one letter between
    let has_repeat = (0..s.len()-2).any(|i| {
        s.as_bytes()[i] == s.as_bytes()[i+2]
    });

    has_repeat
}

fn main() {
    // Test cases with assertions
    assert_eq!(is_nice("qjhvhtzxzqqjkmpb"), true);
    assert_eq!(is_nice("xxyxx"), true);
    assert_eq!(is_nice("uurcxstgmygtbstg"), false);
    assert_eq!(is_nice("ieodomkazucvgmuy"), false);

    // Read input and calculate result
    let content = fs::read_to_string("../input.txt")
        .expect("Failed to read input file");
    
    let count = content.lines().filter(|&s| is_nice(s)).count();
    println!("Rust: {}", count);
}
